ec2_group VPC fixes

willthames opened the pull request:
##### SUMMARY
Fix some of the minor flaws in ec2_group identified since the boto3 improvement.

Add tests for all identified problems.

Massively improve information returned by ec2_group and add RETURN to reflect.

##### ISSUE TYPE
 - Bugfix Pull Request

##### COMPONENT NAME
ec2_group

##### ANSIBLE VERSION

```
ansible 2.4.0 (devel 317ad54599) last updated 2017/07/19 21:52:32 (GMT +1000)
  config file = None
  configured module search path = [u'/Users/will/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /Users/will/src/ansible/lib/ansible
  executable location = /Users/will/src/ansible/bin/ansible
  python version = 2.7.13 (default, Dec 18 2016, 07:03:39) [GCC 4.2.1 Compatible Apple LLVM 8.0.0 (clang-800.0.42.1)]

```


##### ADDITIONAL INFORMATION

Commit:
Add tests for group in a VPC

Commit:
Improve ec2_group output and documentation

Update ec2_group to provide full security group information
Add RETURN documentation to match

Commit:
Fix ec2_group creation within a VPC

Ensure VPC ID gets passed when creating security group

Commit:
Add test for auto creating SG

willthames commented:
@mattclay this will need DeleteVpc permissions

samdoran commented:
cc @s-hertel

Commit:
Fix ec2_group auto group creation

s-hertel commented on lib/ansible/modules/cloud/amazon/ec2_group.py:
```diff
@@ -296,26 +362,29 @@ def get_target_from_rule(module, ec2, rule, name, group, groups, vpc_id):
     elif 'group_id' in rule and re.match(FOREIGN_SECURITY_GROUP_REGEX, rule['group_id']):
```
This got fixed one already but was broken during the boto -> boto3 upgrade and I missed it. Can you make this
`elif rule.get('group_id') and re.match(FOREIGN_SECURITY_GROUP_REGEX, rule['group_id']):` please? Refixes if group_id is null. Otherwise the regex fails:
```
"Traceback (most recent call last):\n  File \"/var/folders/by/k8_fbl593dlctgqmwq5wzl2c0000gn/T/ansible_2srlf662/ansible_module_ec2_group.py\", line 836, in <module>\n    main()\n  File \"/var/folders/by/k8_fbl593dlctgqmwq5wzl2c0000gn/T/ansible_2srlf662/ansible_module_ec2_group.py\", line 680, in main\n    group, groups, vpc_id)\n  File \"/var/folders/by/k8_fbl593dlctgqmwq5wzl2c0000gn/T/ansible_2srlf662/ansible_module_ec2_group.py\", line 362, in get_target_from_rule\n    elif 'group_id' in rule and re.match(FOREIGN_SECURITY_GROUP_REGEX, rule['group_id']):\n  File \"/Users/shertel/Workspace/ansible/venv/python3.5.2/lib/python3.5/re.py\", line 163, in match\n    return _compile(pattern, flags).match(string)\nTypeError: expected string or bytes-like object\n",
```
Though a follow up pr is fine too.

willthames replied:
No, I'll fix this now

s-hertel commented on test/integration/targets/ec2_group/tasks/main.yml:
```diff
@@ -343,10 +368,45 @@
+        rules:
+        - proto: "tcp"
+          from_port: 8182
+          to_port: 8182
```
It appears between 2.3 and now we lost the functionality provided by https://github.com/ansible/ansible/pull/20806. I suspect that also happened during the boto -> boto3 transition. Let me know if you want to do that or if you want me to make a follow up pr. Either is fine, but that would be another thing we need a test for/need to re-fix.

s-hertel commented on test/integration/targets/ec2_group/tasks/main.yml:
```diff
@@ -325,6 +325,31 @@
           - 'not result.changed'
           - 'result.group_id.startswith("sg-")'
 
+    - name: add a rule that auto creates another security group
```
Hurrah for more tests!

willthames replied:
Indeed - as I find issues, I'm doing TDD - add a test, make sure it fails, fix the code, make sure the test passes.

mattclay replied:
Please include cleanup for this security group in the always block.

mattclay replied:
Also, when possible, `resource_prefix` should be included in the names of resources so we can more easily identify which test run they came from, should they survive beyond the test run.

s-hertel reviewed (approved):
The tests are passing locally and this looks pretty straightforward.

Commit:
Add backoff to describe_security_groups

Getting LimitExceeded from describe_security_groups is definitely
possible (source: me) so add backoff to increase likelihood of
success.

To ensure that all `describe_security_group` calls are backed off,
remove implicit ones that use `ec2.SecurityGroup`. From there,
the decision to remove the `ec2` boto3 resource and rely on the client
alone makes good sense.

s-hertel commented:
@willthames Looks like there are conflicts now :/ Can you rebase?

@mattclay Can you add permissions please?

s-hertel commented:
Now needing permission for ec2:DescribeVpcs. I guess the wildcard doesn't work for that? :/

mattclay commented:
The test policy has been updated. Tests are now passing.

mattclay reviewed (changes requested):
Thanks for the fixes, additional documentation and tests. It looks good overall. Just two minor changes to the tests requested (see previous comments).

Commit:
Tidy up auto created security group

Add resource_prefix to auto created security group and delete
it in the `always` section.
Use YAML argument form for all module parameters

willthames commented:
@s-hertel the wildcard should indeed work - I haven't had any problems running the tests with just the included policies - can you let me know what error you're seeing? Probably worth checking your IAM role does include the `Describe*` - I had some problems with Default policy versions in my own tests just now that meant it look like I had DeleteVpc but the default policy didn't, because I'd been flipping back and forth between another set of tests with a different set of IAM policies.

willthames commented:
@mattclay your concerns should now be addressed.

mattclay commented:
@willthames Yes, thank you. Please note that I haven't reviewed the whole PR, I just spotted the security group naming and cleanup while testing the IAM policy updates.
